ocaml/xapi-storage-script/lib.ml

```diff
@@ -0,0 +1,259 @@
+              Lwt.return Regular
+          | Unix.S_DIR ->
+              Lwt.return Directory
+          | _ ->
+              Lwt.return Other
+        )
+        )
+      (function
+        | Unix.Unix_error (Unix.ENOENT, _, _) ->
+            Lwt.return Missing
+        | Unix.Unix_error ((Unix.EACCES | Unix.ELOOP), _, _) ->
+            Lwt.return Unknown
+        | e ->
+            Lwt.fail e
+        )
+
+  let access path modes =
+    Lwt.try_bind
+      (fun () -> Lwt_unix.access path modes)
+      Lwt_result.return
+      (fun exn -> fail (`not_executable (path, exn)))
+
+  let assert_is_executable path =
+    file_kind ~follow_symlinks:true path >>= function
+    | Directory | Other | Missing | Unknown ->
+        fail (`missing path)
+    | Regular -> (
+        access path [Unix.X_OK] >>= function
+        | Error exn ->
+            fail exn
+        | Ok () ->
+            return ()
+      )
+
+  let read_file_contents path =
+    Lwt_io.(with_file ~mode:input ~flags:[O_RDONLY] ~perm:0o000 path read)
+
+  let save ~contents path =
+    Lwt_io.(with_file ~mode:output path (Fun.flip write contents))
+
+  let readdir path =
+    path |> Lwt_unix.files_of_directory |> Lwt_stream.to_list >>= fun listing ->
+    List.filter (function "." | ".." -> false | _ -> true) listing
+    |> Lwt.return
+
+  let mkdir_p ?(perm = 0o755) path =
+    let rec loop acc path =
+      let create_dir () = Lwt_unix.mkdir path perm in
+      let create_subdirs () = Lwt_list.iter_s (fun f -> f ()) acc in
+      Lwt.try_bind create_dir create_subdirs (function
+        | Unix.(Unix_error (EEXIST, _, _)) ->
+            (* create directories, parents first *)
+            create_subdirs ()
+        | Unix.(Unix_error (ENOENT, _, _)) ->
+            let parent = Filename.dirname path in
+            loop (create_dir :: acc) parent
+        | exn ->
+            let msg =
+              Printf.sprintf {|Could not create directory "%s" because: %s|}
+                path (Printexc.to_string exn)
+            in
+            Lwt.fail (Failure msg)
+        )
+    in
+    loop [] path
+end
+
+module Signal = struct
+  type t = int
+
+  let to_string s = Fmt.(str "%a" Dump.signal s)
+end
+
+module Process = struct
+  module Output = struct
+    type exit_or_signal = Exit_non_zero of int | Signal of Signal.t
+
+    type t = {
+        exit_status: (unit, exit_or_signal) Result.t
+      ; stdout: string
+      ; stderr: string
+    }
+
+    let exit_or_signal_of_unix = function
+      | Unix.WEXITED 0 ->
+          Ok ()
+      | WEXITED n ->
+          Error (Exit_non_zero n)
+      | WSIGNALED n ->
+          Error (Signal n)
+      | WSTOPPED n ->
+          Error (Signal n)
+  end
+
+  let create ~env ~prog ~args =
+    let args = Array.of_list (prog :: args) in
+    let cmd = (prog, args) in
+
+    let env =
+      Unix.environment ()
+      |> Array.to_seq
+      |> Seq.map (fun kv ->
+             let k, v = Scanf.sscanf kv "%s@=%s" (fun k v -> (k, v)) in
```

Comment:
This can fail with an exception that is not caught; however, `Unix.environment` should guarantee what we find this format.

Reply:
Reading the POSIX manual here is interesting. It is very careful with its wording, and says that by convention the environment has `key=value` pairs.
But nothing prevents you from doing this, in which case you don't get `key=value` pairs:
```
utop# Unix.create_process_env "./printenv" [||] [|"FOO"; "BAR=4"|] Unix.stdin Unix.stdout Unix.stderr;;
FOO
BAR=4
```

This is consistent with how `/usr/bin/env` works it also prints the environment without any sanitization or checks:
```
utop# Unix.create_process_env "/usr/bin/env" [||] [|"FOO"; "BAR=4"|] Unix.stdin Unix.stdout Unix.stderr;;
FOO
BAR=4
```

Python has a dictionary for `os.environ` and apparently it filters out environment variables that don't match the convention completely:
```
utop# Unix.create_process_env "/usr/bin/python3" [|"python3"; "-c"; "import os\nprint(os.environ)"|] [|"FOO";"BAR=4"|] Unix.stdin Unix.stdout Unix.stderr;;
 environ({'BAR': '4', 'LC_CTYPE': 'C.UTF-8'})
```

Also note that `putenv` has a non-standard extension in Linux where calling `putenv("FOO")` removes `FOO` from the environment.

Probably not erelvant in our case (since we are likely the ones calling these executables in the first place), but if we want to be fully generic then we can filter out or reject environment entries that are not key=value pairs.

I'd go with reject, that is a very unusual situation that is quite difficult to create (unless you deliberately call `execve` like I've done above), and probably an indication of a bug somewhere.